wasm/derive/src/validator/entrypoint.rs

```diff
@@ -0,0 +1,160 @@
+            use ::alloc::format;
+
+            let needs_permission = ::iroha_wasm::query_operation_to_validate();
+            <
+                ::iroha_wasm::data_model::prelude::#operation_type
+                as ::core::convert::TryFrom<
+                    ::iroha_wasm::data_model::permission::validator::NeedsPermissionBox
+                >
+            >::try_from(needs_permission)
+                .dbg_expect(&format!(
+                    "Failed to convert `NeedsPermissionBox` to `{}`. \
+        Have you set right permission validator type?",
+                    stringify!(#operation_type)
+                ))
+        }}
+    }
+}
+
+impl crate::params::ConstructArg for ParamType {
+    fn construct_arg(&self) -> syn::Expr {
+        match self {
+            ParamType::Authority => {
+                parse_quote! {
+                    ::iroha_wasm::query_authority()
+                }
+            }
+            ParamType::Transaction => {
+                Self::construct_operation_arg(&parse_quote!(SignedTransaction))
+            }
+            ParamType::Instruction => Self::construct_operation_arg(&parse_quote!(Instruction)),
+            ParamType::Query => Self::construct_operation_arg(&parse_quote!(QueryBox)),
+            ParamType::Expression => Self::construct_operation_arg(&parse_quote!(Expression)),
+        }
+    }
+}
+
+/// [`validator_entrypoint`](crate::validator_entrypoint()) macro implementation
+#[allow(clippy::needless_pass_by_value)]
+pub fn impl_entrypoint(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let syn::ItemFn {
+        attrs,
+        vis,
+        sig,
+        mut block,
+    } = parse_macro_input!(item);
+
+    let fn_name = &sig.ident;
+    assert!(
+        matches!(sig.output, syn::ReturnType::Type(_, _)),
+        "Validator entrypoint must have `Verdict` return type"
+    );
+
+    let args = match syn::parse_macro_input!(attr as Attr) {
+        Attr::Params(params_attr) => {
+            let operation_param_count = params_attr
+                .types()
+                .filter(|param_type| *param_type != &ParamType::Authority)
+                .count();
+            assert!(
+                operation_param_count == 1,
+                "Validator entrypoint macro attribute must have exactly one parameter \
+                of some operation type: `transaction`, `instruction`, `query` or `expression`"
+            );
+
+            params_attr.construct_args()
+        }
+    };
+
+    block.stmts.insert(
+        0,
+        parse_quote!(
+            use ::iroha_wasm::Execute as _;
+        ),
+    );
+
+    quote! {
+        /// Validator entrypoint
+        ///
+        /// # Memory safety
+        ///
+        /// This function transfers the ownership of allocated
+        /// [`Verdict`](::iroha_wasm::data_model::permission::validator::Verdict)
+        #[no_mangle]
+        #[doc(hidden)]
+        pub unsafe extern "C" fn _iroha_validator_main()
```

Comment:
```suggestion
        pub unsafe extern "C" fn _iroha_wasm_main()
```

I get an urge to want to overload it and take care of function prototype it on the host side

Reply:
I thought about it, but have no desire to implement. And it also related to triggers then. Let's move it to another issue

Reply:
Created #3263